app/grandchallenge/core/signals.py

```diff
@@ -54,11 +58,50 @@ def process_permission_request(sender, instance, *_, **__):
     old_status = old_values.status if old_values else None
 
     if instance.status != old_status:
-        if instance.status == instance.PENDING:
-            send_permission_request_email(instance)
-        elif instance.status == instance.ACCEPTED:
+        if instance.status == instance.ACCEPTED:
             instance.add_method(instance.user)
-            send_permission_granted_email(instance)
-        else:
+            action.send(
+                sender=instance, verb="was accepted",
+            )
+        elif instance.status == instance.REJECTED:
             instance.remove_method(instance.user)
-            send_permission_denied_email(instance)
+            action.send(
+                sender=instance, verb="was rejected",
+            )
+
+
+@receiver(post_save, sender=AlgorithmPermissionRequest)
+@receiver(post_save, sender=ArchivePermissionRequest)
+@receiver(post_save, sender=ReaderStudyPermissionRequest)
+def process_permission_request(sender, created, instance, *_, **__):
+    if created:
+        follow(
+            user=instance.user,
+            obj=instance,
+            actor_only=False,
+            send_action=False,
+        )
+        action.send(
+            sender=instance.user,
+            verb="requested access to",
+            target=instance.base_object,
+        )
+
+
+@receiver(post_save, sender=Algorithm)
+@receiver(post_save, sender=Archive)
+@receiver(post_save, sender=ReaderStudy)
+def create_editor_follows(sender, created, instance, *_, **__):
+    if created:
+        for user in instance.editors_group.user_set.all():
+            follow(user=user, obj=instance, send_action=False)
+
+
+@receiver(pre_delete, sender=Algorithm)
+@receiver(pre_delete, sender=Archive)
+@receiver(pre_delete, sender=ReaderStudy)
+@receiver(pre_delete, sender=AlgorithmPermissionRequest)
+@receiver(pre_delete, sender=ArchivePermissionRequest)
+@receiver(pre_delete, sender=ReaderStudyPermissionRequest)
+def delete_editor_follows(sender, instance, *_, **__):
+    Follow.objects.filter(object_id=instance.pk).delete()
```

Comment:
Signals really should only be used as a last resort as it makes code very difficult to reason about. We are in control of these models, and will not have a problem with circular imports, which means that we can extend their `save` and `delete` methods rather than using signals.